h/groups/views.py

```diff
@@ -101,26 +104,28 @@ def _read_group(request, group):
     """
     url = request.route_url('group_read', pubid=group.pubid, slug=group.slug)
 
-    result = search.search(request,
-                           private=False,
-                           params={"group": group.pubid, "limit": 1000})
-    annotations = [presenters.AnnotationHTMLPresenter(
-                       storage.annotation_from_dict(a))
-                   for a in result['rows']]
-
-    # Group the annotations by URI.
-    # Create a dict mapping the (normalized) URIs of the annotated documents
-    # to the most recent annotation of each document.
-    annotations_by_uri = collections.OrderedDict()
-    for annotation in annotations:
-        normalized_uri = uri.normalize(annotation.uri)
-        if normalized_uri not in annotations_by_uri:
-            annotations_by_uri[normalized_uri] = annotation
-            if len(annotations_by_uri) >= 25:
-                break
-
-    document_links = [annotation.document_link
-                      for annotation in annotations_by_uri.values()]
+    # Find documents that have at least one annotation that:
+    # a. is shared, and
+    # b. is in this group.
+    documents = (
+        request.db.query(models.Document)
+        .options(joinedload(models.Document.meta))
+        .options(joinedload(models.Document.document_uris))
+        .filter(
+            models.Document.annotations.any(
+                and_(
+                    models.Annotation.shared == expression.true(),
+                    models.Annotation.groupid == group.pubid
+                )
+            )
+        )
+        .order_by(models.Annotation.created)
+        [:25]
```

Comment:
I'd argue for `[:25]` as more readable, being the standard way to limit a sequence in Python and sqlalchemy (the examples in the sqla tutorial docs use slicing not `limit()`, though of course `limit()` is documented when you get to the API docs). I'm happy to change this if you prefer `limit()` though.